### ISSUE TITLE
Reproduce inductive results using pykeen code

### ISSUE THREAD

**Issue by nomisto** [migalkin/NodePiece#7]
Hello,

can you spot any obvious reasons why I cannot reproduce your numbers for e.g. the fb15k-237_v1 inductive set, with the pykeen code (See below). Any help would be appreciated! (Most of the code is from here https://pykeen.readthedocs.io/en/stable/tutorial/inductive_lp.html#full-inductive-lp-example, however some parameters were adapted, num epochs, lr, gnn_encoder, etc. )

With this code I can only achieve a hits@10 of about 0.7

Thanks

```python
from pykeen.datasets.inductive.ilp_teru import InductiveFB15k237
from pykeen.models.inductive import InductiveNodePieceGNN
from pykeen.training import SLCWATrainingLoop
from pykeen.evaluation.rank_based_evaluator import SampledRankBasedEvaluator
from pykeen.stoppers import EarlyStopper
from torch import nn
from pykeen.nn.representation import CompGCNLayer
from torch.optim import Adam
from pykeen.losses import BCEWithLogitsLoss, NSSALoss

epochs = 2500

dataset = InductiveFB15k237(version="v1", create_inverse_triples=True)
model = InductiveNodePieceGNN(
    triples_factory=dataset.transductive_training,  # training factory, will be also used for a GNN
    inference_factory=dataset.inductive_inference,  # inference factory, will be used for a GNN
    num_tokens=12,  # length of a node hash - how many unique relations per node will be used
    aggregation="mlp",
    loss=NSSALoss(margin=25),  # dummy loss
    random_seed=42,
    gnn_encoder= nn.ModuleList([
        CompGCNLayer(input_dim=100, output_dim=100, dropout=0.1, activation="relu", attention_heads = 1, edge_weighting="AttentionEdgeWeighting") for _ in range(3)
    ]),
    embedding_dim=100,
    interaction="rotate"
)

import wandb
from pykeen.trackers import WANDBResultTracker
project_name = "evals"
tracker = WANDBResultTracker(project=project_name, group=None, settings=wandb.Settings(start_method='fork'))
tracker.start_run()
tracker.wandb.config.update({"test": "inductive_nodepiece_gnn"})

optimizer = Adam(params=model.parameters(), lr=0.0001)

training_loop = SLCWATrainingLoop(
    triples_factory=dataset.transductive_training,  # training triples
    model=model,
    optimizer=optimizer,
    negative_sampler_kwargs=dict(num_negs_per_pos=32),
    mode="training",   # necessary to specify for the inductive mode - training has its own set of nodes
    result_tracker=tracker,
    automatic_memory_optimization=False
)

# Validation and Test evaluators use a restricted protocol ranking against 50 random negatives
valid_evaluator = SampledRankBasedEvaluator(
    mode="validation",   # necessary to specify for the inductive mode - this will use inference nodes
    evaluation_factory=dataset.inductive_validation,  # validation triples to predict
    additional_filter_triples=dataset.inductive_inference.mapped_triples,   # filter out true inference triples
)

# According to the original code
# https://github.com/kkteru/grail/blob/2a3dffa719518e7e6250e355a2fb37cd932de91e/test_ranking.py#L526-L529
# test filtering uses only the inductive_inference split and does not include inductive_validation triples
# If you use the full RankBasedEvaluator, both inductive_inference and inductive_validation triples
# must be added to the additional_filter_triples
test_evaluator = SampledRankBasedEvaluator(
    mode="testing",   # necessary to specify for the inductive mode - this will use inference nodes
    evaluation_factory=dataset.inductive_testing,  # test triples to predict
    additional_filter_triples=dataset.inductive_inference.mapped_triples,   # filter out true inference triples
)

early_stopper = EarlyStopper(
    model=model,
    training_triples_factory=dataset.inductive_inference,
    evaluation_triples_factory=dataset.inductive_validation,
    frequency=1,
    patience=100000,  # for test reasons, turn it off
    evaluation_batch_size=256,
    evaluator=valid_evaluator,
    result_tracker=tracker
)

# Training starts here
training_loop.train(
    triples_factory=dataset.transductive_training,
    stopper=early_stopper,
    num_epochs=epochs,
    batch_size=512,
)

# Test evaluation
result = test_evaluator.evaluate(
    model=model,
    mapped_triples=dataset.inductive_testing.mapped_triples,
    additional_filter_triples=dataset.inductive_inference.mapped_triples,
    batch_size=256,
)

tracker.log_metrics(
    metrics=result.to_flat_dict(),
    step=epochs + 1,
    prefix='test',
)
tracker.end_run(success=True)

# print final results
print(result.to_flat_dict())
```

**Comment by migalkin** [comment]
There has been quite some refactoring of pykeen's internals as to the underlying representations including those with complex numbers. Can't spot obvious problems in the code, but can you try the `distmult` interaction?

**Comment by nomisto** [comment]
When using `distmult` I get similar results, however if I increase the `embedding_dim=200` the `rotate` improves significantly.
|   | test_h@10 | valid_h@10 |
| ------------- | ------------- |------------- |
| `rotate` `embedding_dim=100` |  0.66 |  0.68 |
| `rotate` `embedding_dim=200` | 0.775 | 0.77 |
| `distmult` `embedding_dim=100` | 0.66 | 0.76 |
 |`distmult` `embedding_dim=200` | 0.69 | 0.73 |

Yes I also thought about the complex numbers and if `embeddings_dim` is both real + complex or for each. However pykeen: `embedding_dim=100` and your code `embedding_dim=100` have exactly the same number of parameters.

Things that I noticed:
a) The test h@10 is often lower than the validation. If I am not incorrect, its usually the other way around.
b) Depending on your hardware, but with a 2080 titan my runtimes are 45 minutes for this, yours are about 2 hours
c) Although `distmult` `embedding_dim=100` valid is high, test is quite worse.

Maybe the evalutation process differs? I know that the dataset of v1 is quite small, but 0.2 hits@10 difference is quite significant.

**Comment by migalkin** [comment]
> The test h@10 is often lower than the validation. If I am not incorrect, its usually the other way around.

It's normal and usual that validation performance is better than test

> Depending on your hardware, but with a 2080 titan my runtimes are 45 minutes for this, yours are about 2 hours

There have been many changes in pykeen since 2021 both in terms of training speedup and loss scaling - so the same 2500 epochs in newer versions might not result into the same final states (perhaps in the newer versions you'd need to train longer)

I also found [here](https://github.com/migalkin/NodePiece/blob/9adc57efe302919d017d74fc648f853308cf75fd/inductive_lp/model/nodepiece_rotate.py#L45) that the default number of attention heads is 2 whereas your setup has one head. Try with 2?

https://github.com/migalkin/NodePiece/blob/9adc57efe302919d017d74fc648f853308cf75fd/inductive_lp/model/nodepiece_rotate.py#L45

**Comment by nomisto** [comment]
Ah thanks, I missed that. And you were right, it was the small learning rate. I could now get comparable results with the pykeen code and a learning rate of `0.005`. Thanks for your help!
